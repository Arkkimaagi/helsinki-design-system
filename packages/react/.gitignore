*.css.d.ts
storybook-static