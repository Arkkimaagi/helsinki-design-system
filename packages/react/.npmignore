.*
**/tsconfig.json
**/rollup.config.js
**/jest.config.js
scripts
src
storybook-static
