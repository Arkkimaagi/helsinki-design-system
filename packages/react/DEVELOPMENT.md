# Development

## Development environment

> Helsinki Design System uses [**Lerna**](https://lerna.js.org/) for running scripts across the repo as well as versioning and creating releases of the packages. [**Yarn workspaces**](https://yarnpkg.com/lang/en/docs/workspaces/) is used to manage dependencies. This allows the separate packages to reference each other via symlinks during local development.

### Setting up local development environment

1. Clone the HDS repository.
```bash
git clone https://github.com/City-of-Helsinki/helsinki-design-system.git
```

2. Go to the root of the project and install dependencies with `yarn`.
```bash
cd helsinki-design-system
yarn
```

3. Build the hds-core and hds-design-tokens packages. hds-react uses these packages as dependencies.
```bash
yarn build:tokens && yarn build:core
```

4. Start the development server.
```bash
yarn start:react
```

This starts the storybook development environment. You can read about how to write storybook stories [here](https://storybook.js.org/docs/react/get-started/whats-a-story). The use of Storybook is recommended for component development.

### Commands                                                                                  
                                                                                              
| Command                            | Description                                               | 
| ---------------------------------- | --------------------------------------------------------- | 
| yarn                               | Install dependencies and link local packages.             | 
| yarn build                         | Builds the package.                                       | 
| yarn start                         | Starts the development environment.                       | 
| yarn test                          | Runs the tests.                                           | 
| yarn lint                          | Runs the linting.                                         | 
| yarn scaffold                      | Runs the scaffolding script for creating a new component. | 

## Adding a component

To add a new component, use the scaffold helper. In the react package folder, run:

```
yarn scaffold
```

## Workflow

This project uses the [Git Feature Branch Workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow). Happy branching!

## Git

Pull requests can be submitted from fork. [Read more from here.](https://guides.github.com/activities/forking/)

### Developing a feature:

```bash
git checkout -b <branchname>
git add <file1> <file2> ...
git commit -m "My commit message"
```

### Pushing your feature to Github:

```bash
git checkout develop
git pull
git checkout <branchname>
git rebase -i develop
```

- Resolve conflicts and continue:

```bash
git add <file1> <file2> ...
git rebase --continue
```

- After no conflicts:

```bash
git push --force-with-lease
```

(If your remote does not accept your local new branch: `git push -u origin HEAD`)

- Make a Pull Request at Github website.
