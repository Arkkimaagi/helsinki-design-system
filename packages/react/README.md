# hds-react

A collection of Helsinki Design System components implemented using React.

## Getting started

> **Helsinki Design System** uses [**Lerna**](https://lerna.js.org/) and [**Yarn workspaces**](https://yarnpkg.com/lang/en/docs/workspaces/) to manage dependencies during local development. This allows the separate packages to reference each other locally via symlinks.

```
# Clone the repo
git clone https://github.com/City-of-Helsinki/helsinki-design-system.git
cd helsinki-design-system

# Install dependencies
yarn

# Start React-development (watch packages for changes, launch Storybook)
yarn start:react
```

## Built with

- React with [TypeScript](https://www.typescriptlang.org/) support
- bundled with [Webpack](https://github.com/webpack/webpack), compiled with [Babel](https://github.com/babel/babel)
- [CSS Modules](https://github.com/css-modules/css-modules) with [typed-css-modules](https://github.com/Quramy/typed-css-modules) for styles
- [ESLint](https://github.com/eslint/eslint) for code analysis
- [Jest](https://github.com/facebook/jest) with [React Testing Library](https://github.com/testing-library/react-testing-library) for testing

## Contributing

The initial version of the Helsinki Design System is under development. We will accept new features, feature requests and help with improving the documentation in the future.
https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow

    "build": "yarn run lint && yarn run tcm && yarn run tsc && webpack --mode production",
    "start": "webpack --watch --mode development & tcm src --watch & yarn storybook",
    "tsc": "tsc",
    "tcm": "tcm src",
    "lint": "yarn eslint \"src/**/*.{ts,tsx}\"",
    "pre-commit": "yarn -s lint",
    "pre-push": "yarn -s test",
    "storybook": "start-storybook -p 6006 --quiet",
    "build-storybook": "build-storybook",
    "test": "jest"
