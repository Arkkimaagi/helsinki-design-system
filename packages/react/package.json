{
  "name": "hds-react",
  "version": "1.0.0",
  "license": "MIT",
  "main": "lib/index.js",
  "scripts": {
    "build": "yarn run tcm && yarn run tsc && webpack --mode production",
    "watch": "webpack --watch --mode development & tcm src --watch",
    "tsc": "tsc",
    "tcm": "tcm src"
  },
  "devDependencies": {
    "@babel/core": "^7.6.2",
    "@babel/preset-env": "^7.6.2",
    "@babel/preset-react": "^7.0.0",
    "@babel/preset-typescript": "^7.6.0",
    "@types/react": "^16.9.5",
    "@types/react-dom": "^16.9.1",
    "babel-loader": "^8.0.6",
    "css-loader": "^3.2.0",
    "style-loader": "^1.0.0",
    "typed-css-modules": "^0.6.2",
    "typescript": "^3.6.3",
    "webpack": "^4.41.0",
    "webpack-cli": "^3.3.9"
  },
  "dependencies": {
    "react": "16.8.5",
    "react-dom": "16.8.5",
    "hds-core": "^1.0.0"
  },
  "peerDependencies": {
    "react": "^15.0.0 || ^16.0.0",
    "react-dom": "^15.0.0 || ^16.0.0"
  }
}
