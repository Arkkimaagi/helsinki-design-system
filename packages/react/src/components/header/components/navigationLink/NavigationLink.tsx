import React, { MouseEventHandler, cloneElement, useCallback, useEffect, useRef, useState } from 'react';
// import base styles
import '../../../../styles/base.css';
import { v4 as uuidv4 } from 'uuid';

// import core base styles
import 'hds-core';
import styles from './NavigationLink.module.scss';
import { styleBoundClassNames } from '../../../../utils/classNames';
import { Link } from '../../../link';
import { NavigationLinkDropdown, NavigationLinkInteraction, DropdownMenuPosition } from './navigationLinkDropdown';
import { useHeaderNavigationMenuContext } from '../headerNavigationMenu/HeaderNavigationMenuContext';
import { useHeaderContext } from '../../HeaderContext';
import { MergeElementProps } from '../../../../common/types';

const classNames = styleBoundClassNames(styles);

export type LinkProps = {
  /**
   * Indicator for active link. This is used in HeaderNavigationMenu.
   */
  active?: boolean;
  /**
   * Additional class names to apply.
   */
  className?: string;
  /**
   * Boolean for indicating whether the link has a dropdown menu.
   */
  hasDropdownLinks?: boolean;
  /**
   * Hypertext Reference of the link.
   */
  href: string;
  /**
   * Boolean for indicating whether the link's dropdown menu is open.
   */
  isDropdownOpen?: boolean;
  /**
   * Label for link.
   */
  label: string;
  /**
   * Optional event handler for onMouseEnter.
   */
  onMouseEnter?: MouseEventHandler;
  /**
   * Depth in nested dropdowns.
   * @internal
   */
  depth: number;
};

export type NavigationLinkProps<ReactElement> = {
  /**
   * Indicator for active link. This is used in HeaderNavigationMenu.
   */
  active?: boolean;
  /**
   * Element type to use instead of the default HDS Link.
   * @default Link
   * @example
   * ```ts
   * as={CustomLink}
   * ```
   */
  as?: ReactElement;
  /**
   * Additional class names to apply for the link element.
   */
  className?: string;
  /**
   * Aria-label for the dropdown button to describe closing the dropdown.
   */
  closeDropdownAriaButtonLabel?: string;
  /**
   * Depth in nested dropdowns.
   * @internal
   */
  depth?: number;
  /**
   * Additional class name for the dropdown element.
   */
  dropdownClassName?: string;
  /**
   * Additional class name for the dropdown items.
   */
  dropdownLinkClassName?: string;
  /**
   * Array of NavigationLink components to render in a dropdown. Can be used only inside navigation components.
   */
  dropdownLinks?: Array<React.ReactElement>;
  /**
   * Hypertext Reference of the link.
   * @default #
   */
  href?: string;
  /**
   * Element index given by parent mapping.
   * @internal
   */
  index?: number;
  /**
   * Label for link.
   */
  label: string;
  /**
   * Aria-label for the dropdown button to describe opening the dropdown.
   */
  openDropdownAriaButtonLabel?: string;
  /**
   * Which sub navigation index is open.
   * @internal
   */
  openSubNavIndex?: number;
  /**
   * Set which sub navigation index is open.
   * @internal
   */
  setOpenSubNavIndex?: (val: number) => void;
  /**
   * Additional class name for the dropdown wrapper element.
   */
  wrapperClassName?: string;
};

export type HeaderNavigationLinkProps<ReactElement extends React.ElementType = 'a'> = MergeElementProps<
  ReactElement,
  NavigationLinkProps<ReactElement>
>;

export const NavigationLink = <T extends React.ElementType = 'a'>({
  active,
  as: LinkComponent,
  className,
  wrapperClassName,
  dropdownClassName,
  dropdownLinks,
  dropdownLinkClassName,
  href,
  index,
  label,
  openSubNavIndex,
  setOpenSubNavIndex,
  depth = 0,
  openDropdownAriaButtonLabel,
  closeDropdownAriaButtonLabel,
  ...rest
}: HeaderNavigationLinkProps<T>) => {
  const Item = React.isValidElement(LinkComponent) ? LinkComponent.type : LinkComponent;
  const { isNotLargeScreen } = useHeaderContext();
  const [isDropdownOpen, setDropdownOpen] = useState<boolean>(false);
  const [dropdownOpenedBy, setDropdownOpenedBy] = useState<null | NavigationLinkInteraction>(null);
  const [dynamicPosition, setDynamicPosition] = useState<null | DropdownMenuPosition>(null);
  const { openMainNavIndex, setOpenMainNavIndex } = useHeaderNavigationMenuContext();
  const containerRef = useRef<HTMLSpanElement>(null);
  const isSubNavLink = openSubNavIndex !== undefined && setOpenSubNavIndex !== undefined;

  const handleDynamicMenuPosition = (val: boolean) => {
    // Null position when false so if user resizes browser, the calculation is done again
    if (!val) setDynamicPosition(null);
    // In case of SSR just set the menu to the right
    else if (val && window === undefined) setDynamicPosition(DropdownMenuPosition.Right);
    else {
      // eslint-disable-next-line no-lonely-if
      if (containerRef.current) {
        // Calculate which side has more space for the menu
        const { x: leftPosition, width } = containerRef.current.getBoundingClientRect();
        const rightPosition = leftPosition + width;
        const position =
          leftPosition > window.innerWidth - rightPosition ? DropdownMenuPosition.Left : DropdownMenuPosition.Right;
        setDynamicPosition(position);
      }
    }
  };

  // Handle dropdown open state by calling either internal state or context
  const handleDropdownOpen = (val: boolean, interaction?: NavigationLinkInteraction) => {
    // Set menu position on nested dropdowns to right or left depending on screen size
    if (depth >= 1) handleDynamicMenuPosition(val);
    setDropdownOpenedBy(!val ? null : interaction);
    setDropdownOpen(val);
    // If sub navigation props given, call them
    if (isSubNavLink && index !== undefined) {
      setOpenSubNavIndex(val ? index : -1);
    }
    // Otherwise it's safe to assume that this link is from main navigation and we can call context
    else {
      // eslint-disable-next-line no-lonely-if
      if (((val !== isDropdownOpen && openMainNavIndex === index) || val) && setOpenMainNavIndex) {
        // If closing dropdown, call context only if this is the open main nav dropdown. No need for checks if opening though.
        setOpenMainNavIndex(val ? index : -1);
      }
    }
  };

  const closeDropdown = () => {
    setDropdownOpen(false);
    setDropdownOpenedBy(null);
  };

  useEffect(() => {
    // If sub navigation index is not provided, we need to react to main nav context changes.
    if (!isSubNavLink) {
      // Since only one navigation link menu should be open, close this one if it's not the one that's open.
      if (openMainNavIndex !== index && isDropdownOpen) {
        closeDropdown();
      }
    }
  }, [openMainNavIndex]);

  useEffect(() => {
    // If sub nav and it's index differs from this, this link's dropdown should close
    if (isSubNavLink && openSubNavIndex !== index && isDropdownOpen) {
      closeDropdown();
    }
  }, [openSubNavIndex]);

  // Close dropdown menu when clicked outside
  const handleOutsideClick = useCallback(
    (e: MouseEvent) => {
      if (isDropdownOpen && containerRef.current && !containerRef.current.contains(e.target as Node)) {
        closeDropdown();
      }
    },
    [isDropdownOpen],
  );

  /* Set event listener only when dropdown open */
  useEffect(() => {
    if (isDropdownOpen) document.addEventListener('click', handleOutsideClick);
    return () => document.removeEventListener('click', handleOutsideClick);
  }, [isDropdownOpen]);

  const navigationWrapperLinkClassName = classNames(
    { isNotLargeScreen },
    styles.navigationLinkWrapper,
    styles[`depth-${depth}`],
    wrapperClassName,
  );
  const navigationLinkClassName = classNames(styles.navigationLink, styles[`depth-${depth}`], className, {
    active,
    isNotLargeScreen,
  });

  return (
    <span
      className={navigationWrapperLinkClassName}
      {...(dropdownLinks &&
        dropdownOpenedBy === NavigationLinkInteraction.Hover && {
          onMouseLeave: () => handleDropdownOpen(false),
        })}
      ref={containerRef}
      {...(dropdownLinks &&
        dropdownOpenedBy !== NavigationLinkInteraction.Click && {
          onMouseEnter: () => handleDropdownOpen(true, NavigationLinkInteraction.Hover),
        })}
    >
      <Item
        className={navigationLinkClassName}
        href={href}
        {...(active && { 'aria-selected': true })}
        {...(Boolean(dropdownLinks) && { 'aria-expanded': isDropdownOpen })}
        {...rest}
      >
        {label}
      </Item>
      {dropdownLinks && (
        <NavigationLinkDropdown
          open={isDropdownOpen}
          setOpen={handleDropdownOpen}
          index={index}
          depth={depth + 1}
          className={dropdownClassName}
          dynamicPosition={dynamicPosition}
          openDropdownAriaButtonLabel={openDropdownAriaButtonLabel}
          closeDropdownAriaButtonLabel={closeDropdownAriaButtonLabel}
        >
          {dropdownLinks.map((child) => {
            return cloneElement(child as React.ReactElement, {
              key: uuidv4(),
              wrapperClassName,
              dropdownClassName,
              dropdownLinkClassName,
            });
          })}
        </NavigationLinkDropdown>
      )}
    </span>
  );
};

NavigationLink.defaultProps = {
  as: Link,
  href: '#',
};
