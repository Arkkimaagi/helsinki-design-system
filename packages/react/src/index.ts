// import core base styles
// import 'hds-core';

export * from './icons';
export * from './components';
