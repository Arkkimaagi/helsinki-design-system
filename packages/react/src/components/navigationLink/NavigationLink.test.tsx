import React from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import { NavigationLink } from './NavigationLink';

describe('<NavigationLink /> spec', () => {
  it('renders the component', () => {
    const { asFragment } = render(<NavigationLink href="#">Link</NavigationLink>);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should not have basic accessibility issues', async () => {
    const { container } = render(<NavigationLink href="#">Link</NavigationLink>);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
