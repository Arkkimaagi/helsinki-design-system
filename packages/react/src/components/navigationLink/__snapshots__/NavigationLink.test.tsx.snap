// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavigationLink /> spec renders the component 1`] = `
<DocumentFragment>
  <a
    class="link linkM navigationLink"
    href="#"
  >
    Link
  </a>
</DocumentFragment>
`;
